### PR TITLE
Make exec more consistent with the rest of the kubectl commands.

### LINKED ISSUE #8448: kubectl exec is dissonant with other commands
@smarterclayton @ghodss @bgrant0607 @jlowdermilk 

consider:

```kubectl log <pod> <container>```

```kubectl describe pods <pod>```

vs

```kubectl exec -p <pod> -c <container>```

I propose changing to:

```kubectl exec <pod> <container>```

for continuity.

### PR DESCRIPTION
**Pull request by brendandburns**
Closes https://github.com/GoogleCloudPlatform/kubernetes/issues/8448

### CODE REVIEW

**Comment by dchen1107** [comment]
@brendandburns Run hack/run-gendocs.sh first?

**Comment by brendandburns** [comment]
Updated.  Sorry about that @dchen1107, I don't know why my github prechecks didn't catch that.

--brendan

File: pkg/kubectl/cmd/exec.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ $ kubectl exec -p 123456-7890 -c ruby-container -i -t -- bash -il`
 func NewCmdExec(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *cobra.Command {
 	params := &execParams{}
 	cmd := &cobra.Command{
-		Use:     "exec -p POD -c CONTAINER -- COMMAND [args...]",
+		Use:     "exec POD CONTAINER -- COMMAND [args...]",
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/exec.go]: Please update the examples above.

**Comment by bgrant0607** [comment]
Do a recursive egrep for "kubectl.* exec". A cursory github search found more examples and comments that need to be updated. Also, do we have any e2e tests of exec functionality?

cc @vishh @jlowdermilk @deads2k

File: pkg/kubectl/cmd/exec.go (hunk 1 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: done.

**Comment by brendandburns** [comment]
I found one other example of ```kubectl exec ...``` and I fixed that.  Search for kubectl exec didn't show any other examples.

(and no, I don't think there is an e2e for this [there probably should be])

--brendan

File: pkg/kubectl/cmd/exec.go [hunk 2]
```diff
@@ -32,17 +32,20 @@ import (
 )
 
 const (
-	exec_example = `// get output from running 'date' in ruby-container from pod 123456-7890
-$ kubectl exec -p 123456-7890 -c ruby-container date
+	exec_example = `// get output from running 'date' from pod 123456-7890, using the first container by default
+$ kubectl exec 123456-7890 date
+	
+// get output from running 'date' in ruby-container from pod 123456-7890
+$ kubectl exec 123456-7890 -c ruby-container date
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/exec.go]: Drop the `-c` if the preferred syntax omits it.

File: pkg/kubectl/cmd/exec.go [hunk 3]
```diff
@@ -32,17 +32,20 @@ import (
...
 
 //switch to raw terminal mode, sends stdin to 'bash' in ruby-container from pod 123456-780 and sends stdout/stderr from 'bash' back to the client
-$ kubectl exec -p 123456-7890 -c ruby-container -i -t -- bash -il`
+$ kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il`
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/exec.go]: Ditto

File: pkg/kubectl/cmd/exec.go [hunk 4]
```diff
@@ -32,17 +32,20 @@ import (
...
 )
 
 func NewCmdExec(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer) *cobra.Command {
 	params := &execParams{}
 	cmd := &cobra.Command{
-		Use:     "exec -p POD -c CONTAINER -- COMMAND [args...]",
+		Use:     "exec POD -c CONTAINER -- COMMAND [args...]",
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/exec.go]: Ditto

File: pkg/kubectl/cmd/exec.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/kubectl/cmd/exec.go]: Actually, which is the preferred syntax now, -c or no -c?

File: pkg/kubelet/dockertools/manager.go [hunk 5]
```diff
@@ -1023,7 +1023,7 @@ func (dm *DockerManager) ExecInContainer(containerId string, cmd []string, stdin
 	} else {
 		if stdin != nil {
 			// Use an os.Pipe here as it returns true *os.File objects.
-			// This way, if you run 'kubectl exec -p <pod> -i bash' (no tty) and type 'exit',
+			// This way, if you run 'kubectl exec <pod> -i bash' (no tty) and type 'exit',
```
Comment (bgrant0607) [inline pkg/kubelet/dockertools/manager.go]: Same line appears in pkg/kubelet/rkt/rkt.go

File: pkg/kubectl/cmd/exec.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: I mentioned this in #8448:

I'd prefer to drop the ```-c``` but the cobra parsing makes it impossible to tell the difference between:

```kubectl exec <pod> <container> -- <cmd>```
and
```kubectl exec <pod> --- <cmd> <arg>```

So I think we have to keep the ```-c``` for now.

File: pkg/kubectl/cmd/exec.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: as above.

File: pkg/kubectl/cmd/exec.go (hunk 4 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: as above.

File: pkg/kubelet/dockertools/manager.go (hunk 5 above)
Comment (brendandburns) [inline pkg/kubelet/dockertools/manager.go]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by bgrant0607** [comment]
LGTM

**Comment by thockin** [comment]
needs rebase

**Comment by brendandburns** [comment]
rebased.

**Comment by ncdc** [comment]
@brendandburns FYI I tested this locally (in master) and if you don't specify `-p pod` and instead just do `pod`, it doesn't strip off the 1st command line arg, so

```
kubectl exec nginx date
```

sends the command `nginx date` to the container, whereas

```
kubectl exec -p nginx date
```

correctly sends just `date`

**Comment by ncdc** [comment]
And re e2e tests, I wrote them and left them commented out a while ago, because we didn't have nsenter on the GCE nodes at the time. But once #8864 goes in, we can reenable them (and probably update the code a bit).

**Comment by bgrant0607** [comment]
Could we unit-test the command/arg array without a full-blown e2e test?

Is an issue filed for this yet?

**Comment by ncdc** [comment]
Yes I think it can be unit tested. No I haven't filed an issue yet - I'm
afk now but can do so in a few hours.

**Pull request merged**